text/0000-return-position-impl-trait-in-traits.md

```diff
@@ -0,0 +1,508 @@
+---
+tags: draft-rfc
+title: Return position `impl Trait` in traits
+---
+
+- Feature Name: return_position_impl_trait_in_traits
+- Start Date: 2023-04-27
+- RFC PR: [rust-lang/rfcs#3193](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+- Initiative: [impl trait initiative](https://github.com/rust-lang/impl-trait-initiative)
+
+# Summary
+[summary]: #summary
+
+* Permit `impl Trait` in fn return position within traits and trait impls.
+* Allow `async fn` in traits and trait impls to be used interchangeably with its equivalent `impl Trait` desugaring.
+* Allow trait impls to `#[refine]` an `impl Trait` return type with added bounds or a concrete type.[^refine]
+
+# Motivation
+[motivation]: #motivation
+
+The `impl Trait` syntax is currently accepted in a variety of places within the Rust language to mean "some type that implements `Trait`" (for an overview, see the [explainer] from the impl trait initiative). For function arguments, `impl Trait` is [equivalent to a generic parameter][apit] and it is accepted in all kinds of functions (free functions, inherent impls, traits, and trait impls).
+
+In return position, `impl Trait` [corresponds to an opaque type whose value is inferred][rpit]. In that role, it is currently accepted only in free functions and inherent impls. This RFC extends the support for return position `impl Trait` in functions in traits and trait impls.
+
+[explainer]: https://rust-lang.github.io/impl-trait-initiative/explainer.html
+[apit]: https://rust-lang.github.io/impl-trait-initiative/explainer/apit.html
+[rpit]: https://rust-lang.github.io/impl-trait-initiative/explainer/rpit.html
+
+## Example use case
+
+The use case for `-> impl Trait` in trait functions is similar to its use in other contexts: traits often wish to return "some type" without specifying the exact type. As a simple example that we will use through the RFC, consider the `NewIntoIterator` trait, which is a variant of the existing `IntoIterator` that uses `impl Iterator` as the return type:
```

Comment:
This makes it sound like the primary benefit of RPITIT is ergonomics, rather than adding significant new expressiveness to the language. In other words, it makes some things easier to say but doesn't add the ability to say new things. Is that right?

Reply:
Yes and no. For example, you can do this today:
```rust
trait Foo {
    type Fut<'a>: std::future::Future<Output = ()> + 'a
    where
        Self: 'a;
        
    fn fut(&self) -> Self::Fut<'_>;
}

// Foo::fut is essentially an async fn:
async fn with<F: Foo>(foo: &F) {
    foo.fut().await
}
```
But, as in the RFC, this is not dyn-safe:
```rust
// error[E0191]: the value of the associated type `Fut` (from trait `Foo`) must be specified
async fn with_dyn(foo: &dyn Foo) {
    foo.fut().await
}
```
However, *implementing* `Foo` as written above requires specifying the type of `Foo::Fut` (sometimes annoying, sometimes impossible without type-alias impl-trait), despite the fact that such a definition is entirely redundant with the definition of `Foo::fut`.